File: 198.house-robber/solution.java
class Solution {
    public int rob(int[] nums) {
        int lengthOfNums = nums.length;
        if (lengthOfNums == 1) {
            return nums[0];
        }
        
        int[] dp = new int[lengthOfNums];
        
        dp[0] = nums[0];
        dp[1] = Math.max(dp[0], nums[1]);
        for (int i = 2; i < lengthOfNums; i++) {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i]);
        }
        
        return dp[lengthOfNums - 1];
    }
}

File: 198.house-robber/solution.ts
function rob(nums: number[]): number {

  const moneyState: number[] = new Array(nums.length + 2).fill(0);

  for (let i = 0; i < nums.length; i++) {
      // robbering
      moneyState[i + 2] = moneyState[i] + nums[i];

      // do nothing
      moneyState[i + 1] = Math.max(moneyState[i + 1], moneyState[i]);
  }

  return Math.max(moneyState[nums.length], moneyState[nums.length + 1]);
}